[FEATURE] Implement a QtCreator style locator bar in the QGIS status bar

nyalldawson opened the pull request:
This adds a new "locator" bar to the QGIS status bar. If you're not familiar with QtCreator's locator, it's a quick search bar (activated by Ctrl+K) which displays matching search results from any number of registered search filters.

Search filters are subclassed from QgsLocatorFilter, and added to the app's locator via iface.registerLocatorFilter(...)

Searching is handled using threads, so that results always become available as quickly as possible, regardless of whether any slow search filters may be installed. They also appear as soon as each result is encountered by each filter, which means that e.g. a file search filter will show results one by one
as the file tree is scanned. This ensures that the UI is always responsive even if a very slow search filter is present (e.g. one which uses an online service).

This framework is designed to be extended by plugins, such as OSM nominatim searches, direct database searching (i.e. Discovery plugin), layer catalog searches, etc...

nyalldawson committed:
[FEATURE] Implement a QtCreator style locator bar in the QGIS status bar

This adds a new "locator" bar to the QGIS status bar. If you're not
familiar with QtCreator's locator, it's a quick search bar
(activated by Ctrl+K) which displays matching search results
from any number of registered search filters.

Search filters are subclassed from QgsLocatorFilter, and
added to the app's locator via iface.registerLocatorFilter(...)

Searching is handled using threads, so that results always
become available as quickly as possible, regardless of whether
any slow search filters may be installed. They also appear
as soon as each result is encountered by each filter, which means
that e.g. a file search filter will show results one by one
as the file tree is scanned. This ensures that the UI is always
responsive even if a very slow search filter is present (e.g.
one which uses an online service).

This framework is designed to be extended by plugins, such as
OSM nominatim searches, direct database searching (i.e. Discovery
plugin), layer catalog searches, etc...

nyalldawson commented:
For your review Mr Travis...

nirvn commented:
@nyalldawson , would be great for this locator bar to search for existing composers and open those on name matching.

I.e. I type: "my composer name", and hit enter, *bam* composer opens :)

nyalldawson committed:
Add inbuilt filters for project layers and compositions

The project layer filter allows you to quickly select a layer
from the current project and highlight it in the layer tree.
It's useful for complex project with lots of groups, where
it's easy to "lose" layers somewhere in the tree...

The composition filter allows searching for and opening
compositions from the current project

nyalldawson committed:
Add missing docs, sip bindings

nyalldawson commented:
> @nirvn , would be great for this locator bar to search for existing composers and open those on name matching. I.e. I type: "my composer name", and hit enter, bam composer opens :)

Done!

nirvn commented:
@nyalldawson thanks.

3nids commented:
Do you think this would be appropriate for Quickfinder (plugin) search bar?
The idea would be to bring the results asynchronously and to be able to categorize them.

nyalldawson commented:
> Do you think this would be appropriate for Quickfinder (plugin) search bar?
The idea would be to bring the results asynchronously and to be able to categorize them.

Yes, very much so! It's been designed with exactly this kind of thing in mind.

Some thoughts:
- I'm planning to add a "context" object to QgsLocatorFilter::fetchResults. This could contain things like the current map extent, to prioritise results just like how Quickfinder does it.
- Ideally if you port to this, I'd suggest splitting Quickfinder into multiple smaller plugins. So there'd be OSM search plugin, Geowebfish search plugin and the sqlite based search plugin all as separate self-contained plugins.
- I don't want to use a tree view in the results list. I noticed that Quickfinder does this currently - is that an issue?

nyalldawson committed:
Add a context variable for locator searches

This allows encapsulation of properties relating to the context
of a search, such as a target map extent. Locator filters could
use this to prioritise results close to the current canvas extent.

nyalldawson committed:
Allowing associating QgsLocatorWidget with a map canvas

This allows the widget's locator to prioritise results which
are close to the current canvas extent

nyalldawson commented:
> I'm planning to add a "context" object to QgsLocatorFilter::fetchResults. This could contain things like the current map extent, to prioritise results just like how Quickfinder does it.

Added

3nids commented:
Good idea to use dedicated plug-ins. I'd like to keep a central place for configuration,  probably in general and project settings. 
The tree view is a bit of an issue since it's required to organize results by sources and categories.  Also since results are coming in random order. Why don't you want to go for a tree ?

nyalldawson commented:
> Good idea to use dedicated plug-ins. I'd like to keep a central place for configuration, probably in general and project settings.

I still need to tackle configuration for this (haven't planned anything yet, but realise it'll need doing)

> The tree view is a bit of an issue since it's required to organize results by sources and categories. Also since results are coming in random order. Why don't you want to go for a tree ?

I'm going to group the results with headers for each provider (that's a TODO), and allow filters to specify a custom sort priority for their grouped results too. So that cover half of this (organizing results by source.

I'm against the tree view for a couple of reasons:
1. simplicity from a design point of view, and simplicity for coding of filters
2. tree views have poor keyboard navigation, which is the primary use case here. They push users toward using the mouse to expand/contract nodes. Instead, this is designed to follow the "ctrl +K, then keep typing until you see the result you want, then press up a couple of times to select it and enter" approach.

What kind of categorisation do you desire? I could potentially allow individual filters to have a group for each of their results, so you'd see something like:

    **OSM Search - countries**
    Australia
    Antartica
    **OSM Search - towns**
    Adelaide
    Albury
    **OSM Search - streets**
    ...
     **Bing search - countries**
    ...etc...

But that would be at most one level deep, and I also am reluctant to do this as it adds more non-functional results to the list, making it more likely that good results are hidden off the bottom and require scrolling more often.

3nids commented:
For configuration,  the idea is to provide a virtual configuration widget class to be implemented by the plug-ins? Then it would be nice to have them app-wise and project-based. A new tab in both configuration dialog with a list of added instances with add/remove/edit button ?

For the tree view, one level of depth is perfect as your example illustrates the idea.

NathanW2 commented:
I think the qt creator config screen is good for what we need

On Tue, 9 May 2017, 8:04 PM Denis Rouzaud <notifications@github.com> wrote:

> For configuration, the idea is to provide a virtual configuration widget
> class to be implemented by the plug-ins? Then it would be nice to have them
> app-wise and project-based. A new tab in both configuration dialog with a
> list of added instances with add/remove/edit button ?
>
> For the tree view, one level of depth is perfect as your example
> illustrates the idea.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4512#issuecomment-300119701>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3JtoHsM3BwD-T32M2tU59x-AoCoTks5r4DosgaJpZM4NS_vF>
> .
>

nyalldawson committed:
Add a locator filter for main window actions

Inspired by the ghost of processing's commander

nyalldawson committed:
Add processing algorithms to locator bar

Inspired by the ghost of processing's commander

nyalldawson committed:
Show locator filter names in results, sort results and group by filter

nyalldawson committed:
Allow speciyfing the priority for filters

Higher priority (i.e. more important) filter results get shown
first. This means filters like project layers & composers will
show above 'cruder' filters like the actions/processing filters.

nyalldawson committed:
Implement prefix based locator searching

Filters can indicate their preferred search prefix. Searches which
begin with this character will be restricted to the single matching
filter.

E.g. entering 'l buffer' will searching only layers containing 'buffer'

Other prefixes are:
- . search actions
- pl search print layouts
- a search algorithms

Plugins are restricted to a minimum 3 character prefix. We do this
to avoid plugins 'stealing' desirable prefixes, and instead we
want to reserve them for future core filters.

nyalldawson committed:
Only show action results when . prefix is used

nyalldawson committed:
Sort filter results so closer matches appear higher

A closer match means more of the text is matched, ie
a short string inside a long string is penalised

nyalldawson committed:
Provide a standard string comparison method in QgsLocatorFilter

Subclasses should use this method instead of directly calling
QString::contains or using Python 's in search' type matches.
This ensures consistent matching behaviour across different filters.

nyalldawson committed:
Always select the best result by default, even if good results are
slow to fetch

nyalldawson committed:
Show a list of shortcut prefixes in locator bar if no text is entered

nyalldawson committed:
Nicer results, always show an icon, show description

nyalldawson committed:
Fix QgsFloatingWidget discards some events

nyalldawson committed:
Style fixes

nyalldawson committed:
Add configure menu to locator widget

nyalldawson committed:
Start on options page

nyalldawson committed:
Add (non-functional) locator configuration dialog

nyalldawson committed:
Enable disabling filters via options dialog

nyalldawson committed:
Make locator settings persistent

nyalldawson committed:
Invalidate current locator results when locator options are changed

nyalldawson committed:
Allow filters to provide configuration dialogs

3nids commented on python/gui/locator/qgslocatorfilter.sip:
```diff
@@ -173,6 +173,36 @@ class QgsLocatorFilter : QObject
+ :rtype: bool
+%End
+
+    virtual void openConfigWidget( QWidget *parent = 0 );
```
missing TransferThis?

nyalldawson replied:
No - because that would transfer the filter to that parent. I think this is ok as is - it's up to the subclass implementation to manage the memory here.

3nids replied:
oops, correct.

nyalldawson committed:
Small cleanups

3nids commented on src/app/CMakeLists.txt:
```diff
@@ -519,6 +526,8 @@ INCLUDE_DIRECTORIES(
   openstreetmap
   dwg
   dwg/libdxfrw
+  dwg/libdxfrw
```
listed twice

nyalldawson committed:
Remove duplicate include

nyalldawson committed:
Blacklist PyQgsLocator test on Travis OSX builds

Travis OSX is too underpowered to even run the normal tests...
let alone something that requires... *gasp* multiple cores!!! *gasp*

NathanW2 commented:
Yeeeeeeeeees!

On Thu, 18 May 2017, 7:06 AM Nyall Dawson <notifications@github.com> wrote:

> Merged #4512 <https://github.com/qgis/QGIS/pull/4512>.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4512#event-1086760717>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3AOxbk5DbiVDi_vI5MjiIs0UJBsbks5r62FvgaJpZM4NS_vF>
> .
>

saberraz commented:
My QGIS build on ubuntu 16.04 (Qt 5.6.1) is failing:

[ 47%] Building CXX object src/gui/CMakeFiles/qgis_gui.dir/qgsactionmenu.cpp.o
src/gui/CMakeFiles/qgis_gui.dir/build.make:4818: recipe for target 'src/gui/CMakeFiles/qgis_gui.dir/locator/qgslocatorwidget.cpp.o' failed
make[2]: *** [src/gui/CMakeFiles/qgis_gui.dir/locator/qgslocatorwidget.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 47%] Building CXX object tests/src/core/CMakeFiles/qgis_authconfigtest.dir/qgis_authconfigtest_automoc.cpp.o
Scanning dependencies of target qgis_fieldstest
[ 47%] Building CXX object tests/src/core/CMakeFiles/qgis_fieldstest.dir/testqgsfields.cpp.o
[ 47%] Linking CXX executable ../../../output/bin/qgis_authconfigtest
[ 47%] Building CXX object tests/src/core/CMakeFiles/qgis_fieldstest.dir/qgis_fieldstest_automoc.cpp.o
Scanning dependencies of target qgis_expressioncontexttest
[ 47%] Building CXX object tests/src/core/CMakeFiles/qgis_expressioncontexttest.dir/testqgsexpressioncontext.cpp.o
CMakeFiles/Makefile2:1390: recipe for target 'src/gui/CMakeFiles/qgis_gui.dir/all' failed
make[1]: *** [src/gui/CMakeFiles/qgis_gui.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 47%] Building CXX object tests/src/core/CMakeFiles/qgis_expressioncontexttest.dir/qgis_expressioncontexttest_automoc.cpp.o
[ 47%] Built target qgis_authconfigtest
[ 47%] Linking CXX executable ../../../output/bin/qgis_fieldstest
[ 47%] Linking CXX executable ../../../output/bin/qgis_expressioncontexttest
[ 47%] Built target qgis_fieldstest
[ 47%] Built target qgis_expressioncontexttest
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2

nyalldawson commented:
@saberraz try wiping your build folder and making a clean build

saberraz commented:
@nyalldawson nope...a clean build folder gave me the same error

saberraz commented:
the error was here:
/usr/local/src/QGIS_master/src/gui/locator/qgslocatorwidget.cpp:39:6: error: #elif with no expression

@wonder_sk pointed that it is else not elif

wonder-sk commented:
Fixed now in master
